web/packages/teleport/src/ironrdp/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+        }
+    }
+
+    /// draw_cb: (bitmapFrame: BitmapFrame) => void
+    ///
+    /// respond_cb: (responseFrame: ArrayBuffer) => void
+    pub fn process(
+        &mut self,
+        tdp_fast_path_frame: RDPFastPathPDU,
+        cb_context: &JsValue,
+        draw_cb: &js_sys::Function,
+        respond_cb: &js_sys::Function,
+    ) -> Result<(), JsValue> {
+        let mut output = Vec::new();
+        let tdp_fast_path_frame: RustRDPFastPathPDU = tdp_fast_path_frame.into();
+
+        let graphics_update_region = self
+            .fast_path_processor
+            .process(&mut self.image, &tdp_fast_path_frame.data, &mut output)
+            .map_err(|e| JsValue::from_str(&format!("{:?}", e)))?;
+
+        let mut fast_path_outputs = Vec::new();
+
+        if !output.is_empty() {
+            fast_path_outputs.push(ActiveStageOutput::ResponseFrame(output));
+        }
+
+        if let Some(update_region) = graphics_update_region {
+            fast_path_outputs.push(ActiveStageOutput::GraphicsUpdate(update_region));
+        }
+
+        for out in fast_path_outputs {
+            match out {
+                ActiveStageOutput::GraphicsUpdate(updated_region) => {
+                    // TODO(isaiah): wrap in its own function
+                    let (image_location, image_data) =
+                        extract_partial_image(&self.image, updated_region);
+                    self.apply_image_to_canvas(image_data, image_location, cb_context, draw_cb)?;
+                }
+                ActiveStageOutput::ResponseFrame(frame) => {
+                    // TODO(isaiah): wrap in its own function
+                    let frame = Uint8Array::from(frame.as_slice()); // todo(isaiah): this is a copy
+                    let _ = respond_cb.call1(cb_context, &frame.buffer())?;
+                }
+                ActiveStageOutput::Terminate => {
+                    return Err(JsValue::from_str("Terminate should never be returned"));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn apply_image_to_canvas(
+        &self,
+        image_data: Vec<u8>,
+        image_location: Rectangle,
+        cb_context: &JsValue,
+        callback: &js_sys::Function,
+    ) -> Result<(), JsValue> {
+        let top = image_location.top;
+        let left = image_location.left;
+
+        let image_data = create_image_data_from_image_and_region(&image_data, image_location)?;
+        let bitmap_frame = BitmapFrame {
+            top,
+            left,
+            image_data,
+        };
+
+        let bitmap_frame = &JsValue::from(bitmap_frame);
+
+        // TODO(isaiah): return this?
+        let _ret = callback.call1(cb_context, bitmap_frame)?;
+        Ok(())
+    }
+}
+
+pub fn extract_partial_image(image: &DecodedImage, region: Rectangle) -> (Rectangle, Vec<u8>) {
+    // PERF: needs actual benchmark to find a better heuristic
+    if region.height() > 64 || region.width() > 512 {
```

Comment:
Do you know why it's faster for higher regions? It seems to me that `extract_whole_rows` always copy more data thus should be always slower.